automerge-cli/src/export.rs

```diff
@@ -0,0 +1,42 @@
+use crate::error::AutomergeCliError;
+use automerge_backend::AutomergeError;
+use automerge_frontend::Value;
+use automerge_protocol::Patch;
+use std::fs::File;
+use std::path::Path;
+use std::{io, io::Read};
+
+fn get_changes(changes_file: &Path) -> io::Result<Vec<u8>> {
+    let mut input_data = Vec::new();
+    let mut input_file = File::open(changes_file)?;
+    input_file.read_to_end(&mut input_data)?;
+
+    Ok(input_data)
+}
+
+fn get_patch(changes_bytes: Vec<u8>) -> Result<Patch, AutomergeError> {
+    let mut backend = automerge_backend::Backend::init();
+    backend.load_changes_binary(vec![changes_bytes]).unwrap();
+
+    backend.get_patch()
+}
+
+fn get_state(changes_file: &Path) -> Result<Value, AutomergeCliError> {
+    let input_data =
+        get_changes(changes_file).map_err(|_| AutomergeCliError::InvalidChangesFile)?;
+
+    let patch = get_patch(input_data).map_err(|_| AutomergeCliError::BackendError)?;
```

Comment:
Is using `map_err` idiomatic here to map different error types to application error types?

Reply:
Not really, we do it in a number of places because the codebase is still very rough and ready but I think the idiomatic way to do this is to implement `From<X> for AutomergeCliError>`,  where `X` is the type you are receiving from the fallible call, that allows you to directly use `?` without `map_err`.

[anyhow](https://docs.rs/anyhow/1.0.13/anyhow/) handles a lot of the fiddly parts of this, might be worth a look.

We haven't spent much time sharpening up the error handling for this crate yet, lot's to do there.

Reply:
Idiomatic way is indeed to implement `From<X> for ...`. 

From what I've learned https://docs.rs/anyhow is great for "applciation level" errors, where you want to be able to package and pass along any kind of error.

And for "library level" errors, where there are only a handful of things that can go wrong and you want to expose those for consumers to handle and potentially restart or whatever, there is: https://docs.rs/thiserror instead.

Reply:
Yeah that's my understanding as well. In this case `anyhow` feels appropriate as the CLI is an application consuming the various other crates which are the library.

Reply:
Yeah, I've seen peripherally that error handling has been a big discussion point in Rust lately. I don't have the context to make an informed decision here, so I'm happy to explore `anyhow` here if that's helpful to the overall effort of establishing a pattern! I imagine that library code has different needs than the CLI for error handling, but perhaps it'll turn up some helpful general-purpose discussion?

Reply:
Cool, let's give it a shot and see how it works out.

Reply:
⚡ https://github.com/automerge/automerge-rs/pull/17/commits/89ca6c70b743fa62b556d62c8a0a115d57ef27c8